### DISCUSSION TITLE
Tuning of Gsig for Heterogenous cell size?

### DISCUSSION THREAD

**Discussion by cawarwick** [flatironinstitute/CaImAn#1317] [Q&A]
I have recently started imaging the Dorsal Root Ganglia in which the primary sensory afferents reside. My problem arises from the huge variation in cell body size of these cells in which the largest neurons are >30um in diameter and the smallest are ~10um in diameter. The below image is from a single plane of a 2p recording showing the large variation with a range of around 2-3 fold. I have been messing with gsig and merge_thr to try and find an optimal balance but it feels like a push and pull that I can't get right. e.g. The small cells get merged together with too large of a gsig and so I increase the merge_thr, but then it starts to break up the large cells into 2 pieces. I've done a fair bit of hyperparamter searching and haven't found one I particularly like.

Not sure if there is an optimal solution, but I am not well versed in exactly how CNMF works beyond knowing how critical gsig is to finding cells since it determines the gaussian kernel with which to search for cells.

Thanks!

![image](https://github.com/flatironinstitute/CaImAn/assets/81972652/3372dad8-e89f-4c43-b423-88a13ead14d4)

**Comment by kushalkolar** [comment]
Yea seeding with something besides the greedy init will work better here. If you have a method to segment your cells you can use that as your spatial seeds.

**Reply by cawarwick** [reply]
Ok so it'd be expected that Greedy will struggle with a heterogenous cell size, that's good to know. Would you suggest providing a binary set of masks as is shown in the demo note book?
https://github.com/flatironinstitute/CaImAn/blob/main/demos/notebooks/demo_seeded_CNMF.ipynb

My one thought about this is what while if I provide locations for the seeds (i.e. a mask of the cells), when you start CNMF using those locations as seeds, won't it still rely on gsig to refine the seeds? I guess I just need to try it. I saw an older method from the 2016 paper which suggested "Group Lasso Initialization for Somatic Imaging Data" for densely packed and correlated cells (which these cells certainly are densely packed and often have highly correlated activity). It still uses a gaussian to search but the description of it's application seemed similar to this dataset.  I didn't find any reference to it in the current Caiman so I wasn't sure if it was deprecated or just strictly worse than current initialization methods.

**Reply by kushalkolar** [reply]
Yea I would seed with your own masks. 

What kinds of cells are these? If you're willing to share a snippet of the movie (few thousand frames with activity) I could try an auto initialization algorithm we've been developing. Nowhere near ready for users but curious to test the limits.

**Reply by cawarwick** [reply]
These are first order sensory neurons within the Dorsal Root Ganglia. The recordings are of the soma of these cells. These cells have a psuedo-unipolar anatomy so there are very few axons/dendrites in the ganglia as their axons extend to peripheral tissues (skin/muscle/etc) and the spinal cord, hence why they are so densely packed (and often synchronous) because there is no local circuitry and innervate the same tissues and frequently respond to the same external stimuli. 

All these recordings are a single 2p plane, transgenic expression of Gcamp6s, ~7hz sampling rate. I've been testing gcamp8s as well but don't have many good recordings to share right now.

Short recording with Electrical stimulation (very simultaneous response patterns)
https://pitt-my.sharepoint.com/:i:/g/personal/warwickc_pitt_edu/EYdrVQMGeINBlonVktp9WqYBw8Y7g7cTaYnfy4JU9PMCDg?e=fyXg22

Longer recording with activity that is less synchronous
https://pitt-my.sharepoint.com/:i:/g/personal/warwickc_pitt_edu/EXTZ_TnEtFBJjEu4a5yq-8YB_TIqnzVepKBqwII0DAntzg?e=eWTXf7

**Reply by kushalkolar** [reply]
just downloaded these now, will try to take a look in the coming weeks

**Comment by kushalkolar** [comment]
@cawarwick do you have the video for the image in the first post?

**Reply by cawarwick** [reply]
Here's that exact recoding. These recordings are XYZT and so there are 4 other planes as well if you wanted more data from that same ganglia. I have just included a single plane here:
https://pitt-my.sharepoint.com/:i:/g/personal/warwickc_pitt_edu/EYpaZ0GecZJGv3vr7Ra8k9cBPsnaaE_RURP2llercdLmlg?e=l29YcR

I have also been collecting data using Gcamp8s in these same tissues if that is of any interest

**Comment by kushalkolar** [comment]
Hi, we're making some progress, it's a hard problem especially given the synchronous activity of the cells. Do you have ground truth for any of these movies? If you have more and longer movies that'd be useful too. How long are these videos typically, in number of frames. 

![image-20.png](https://github.com/user-attachments/assets/878ab42d-b603-4eb7-b2fb-58db4bf5237d)

**Reply by cawarwick** [reply]
> Do you have ground truth for any of these movies? 

We do not have any simultaneous patch and imaging, but in some of the recordings we use electrical field stimulation to evoke activity in prescribed patterns (e.g. 1,2,3,4,5,10,20 pulses at 20hz, 2 seconds of pulses at 1,2,5,10,20,40hz etc).

**Reply by kushalkolar** [reply]
By ground truth I don't mean ephys, I mean the segmentation.

Data that's an example of a typical full experiment is best for us to work with.

**Comment by kushalkolar** [comment]
We're getting better seperability, it's still a while before we can have an alpha for users to try out but if you've got more movies, longer is better (how long do you typically acquire for, longer is better):

![image (6)](https://github.com/user-attachments/assets/ca595ca8-a3d7-483b-ad1a-b78055f001d7)

**Reply by cawarwick** [reply]
We've actually got a postdoc who has been doing more of these recordings (especially longer ones). I will check with him and see if he can put together a good recording that has an hour or so of activity. His recordings 'should' have some more varied activity so it might be easier. Is it worth including mulitple Z-planes of the same ganglia or just pick one? e.g. our experiments typically record 4-5 Z-planes using a fast Z-scan so we can get ~400 cells per recording and ~100 per plane.

**Reply by kushalkolar** [reply]
Just one plane of typical data is enough for us to try and figure out how to do this, we're trying to build something that makes no spatial assumptions about neurons which is currently a major limitation with caiman and a nuisance for parameter search.

**Comment by lhastings19** [comment]
I recently started imaging motor neurons, which similarly have a wide range of cell body sizes and a relatively high level of synchronization. I also haven't been able to find parameters that correctly identify the spatial footprints, though I've had some moderate success with seeding with my own masks that I've drawn in ImageJ. I am definitely interested in this pipeline that you all are working on! And I'm happy to share some of my data if that would be helpful at all.

**Reply by kushalkolar** [reply]
Sure, always happy to get more datasets. 

We've hit a roadblock with our approach, it doesn't scale well over ~25 components (~20 components takes 1 minute, 50 components takes overnight) so we're exploring other ideas.

**Reply by lhastings19** [reply]
Sorry for the delay! Here is one of my recordings. I'm using GCaMP8m and the frame rate is 7.5 hz.
https://upenn.box.com/s/4c44d1tgyzz4a129crve4i2192i82fqk

**Reply by lhastings19** [reply]
Hello! Do you have any updates on this effort? I have mainly been doing manual ROI identification but of course I would love to have a better method
